### PR TITLE
Make 3DS launch cores like a normal system

### PR DESCRIPTION
**Pull request by meepingsnesroms**
3dsx build launches .3dsx files, cia build launches .cia files and updates them if they have a different version number.

With this the core system is pretty much crash proof, it loads or it gives an error, no black screen or abort to home menu.

The build system still needs to be updated to place the .cia and .3dsx core files where the .core stubs used to be and needs to give cia builds a version number for updating to work(currently its always 2048/2.0.0).

This also enabled threads by default to fix the lagging on savestate.

### CODE REVIEW

**Comment by fr500** [comment]
is it installing the cias too?

**Comment by meepingsnesroms** [comment]
No it installs them, cias have to be installed to be launched but the library I made handles that on it own and RetroArch doesnt have to do any cia management.

**Comment by aliaspider** [comment]
it installs them of course, that is fine as long as it only happens with cia builds.

**Comment by meepingsnesroms** [comment]
Yes 3dsx is run straight from the file, cia is installed.

**Comment by fr500** [comment]
pretty cool!

**Comment by aliaspider** [comment]
what I meant is that running from a 3dsx shouldn't list the cia files at all, since 3dsx users will not want that.
which is what you are already doing I guess.

File: frontend/drivers/platform_ctr.c [hunk 1]
```diff
@@ -1,564 +1,584 @@
...
+   NULL,                         /* install_signal_handler */
+   NULL,                         /* get_signal_handler_state */
+   NULL,                         /* set_signal_handler_state */
+   NULL,                         /* destroy_signal_handler_state */
+   NULL,                         /* attach_console */
+   NULL,                         /* detach_console */
+   NULL,                         /* watch_path_for_changes */
+   NULL,                         /* check_for_path_changes */
+   "ctr",
+};
```
Comment (aliaspider) [inline frontend/drivers/platform_ctr.c]: what happened with this file ? windows line breaks ? makes it hard to review this way.

**Comment by meepingsnesroms** [comment]
The code didn't match RetroArchs style so I formatted it with artistic style so it wouldn't be a giant mess.

File: frontend/drivers/platform_ctr.c [hunk 2]
```diff
@@ -1,564 +1,584 @@
...
+         exec_3dsx_no_path_in_args(path, args);
+      }
+      else
+      {
+         RARCH_LOG("\n");
+         RARCH_LOG("\n");
+         RARCH_LOG("Warning:\n");
+         RARCH_LOG("First core launch may take 20 seconds!\n");
+         RARCH_LOG("Do not force quit before then or your memory card may be corrupted!\n");
+         RARCH_LOG("\n");
```
Comment (aliaspider) [inline frontend/drivers/platform_ctr.c]: maybe add a progress display during cia install ?
you could also add a prompt for `install`/`cancel` but that is not necessary.

**Comment by aliaspider** [comment]
format changes should be made in a separate PR, otherwise no one can review it.

File: ctr/exec-3dsx/exec_cia.c [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+
+typedef struct{
+	u32 argc;
+	char args[0x300 - 0x4];
+}ciaParam;
+
+
+char argvHmac[0x20] = {0x1d, 0x78, 0xff, 0xb9, 0xc5, 0xbc, 0x78, 0xb7, 0xac, 0x29, 0x1d, 0x3e, 0x16, 0xd0, 0xcf, 0x53, 0xef, 0x12, 0x58, 0x83, 0xb6, 0x9e, 0x2f, 0x79, 0x47, 0xf9, 0x35, 0x61, 0xeb, 0x50, 0xd7, 0x67};
+
```
Comment (aliaspider) [inline ctr/exec-3dsx/exec_cia.c]: doesn't this already exist as an extern ?

**Comment by meepingsnesroms** [comment]
Yes its an extern but my library wount compile outside of retroarch without it and I wanted it to be usable in other projects.

File: ctr/exec-3dsx/exec_cia.c [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+		
+		param.argc = 0;
+		argsLength = 0;
+		if(args != NULL && args[0] != '\0'){
+			bool inSingleQuotes = false;
+			bool inDoubleQuotes = false;
+			int argStringLength = strlen(args);
+			
+			//build argument list like a terminal command on linux
+			for(unsigned int argPtr = 0; argPtr <= argStringLength; argPtr++){
```
Comment (aliaspider) [inline ctr/exec-3dsx/exec_cia.c]: why this change ? there was already a simpler system in place, put argc first then the null terminated arguments after each other in the argv buffer, which is parsed on entry.

**Comment by aliaspider** [comment]
please reply directly under the review comment, and not here ...

File: ctr/exec-3dsx/exec_cia.c (hunk 3 above)
Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: ok

File: ctr/exec-3dsx/exec_cia.c (hunk 4 above)
Comment (meepingsnesroms) [reply ctr/exec-3dsx/exec_cia.c]: Because I wanted it to be like a terminal on Linux so it can be used in other projects.

Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: ??
this is C though not a Linux terminal.
you only need quotes on terminals because you can't input the null character.
you're complicating things by having to add quotes and then having to remove them for no gain.

Comment (meepingsnesroms) [reply ctr/exec-3dsx/exec_cia.c]: In c you can't use the null character either or it will terminate the string, and I wanted a single string argument list.

Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: that is why you put argc first in the buffer ....
the buffer has a fixed size so it will always be copied full.

Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: ok I assumed you were copying the buffer with quotes before launch and parsing them on entry.
you're not doing that , so you can get rid of the quote thing then no ? it makes no sense do to something then then directly call a function that reverts it.

Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: if you want your lib to be able to handle quotes strings as well, add a wrapper for that, but call the simpler function directly from RA.

Comment (meepingsnesroms) [reply ctr/exec-3dsx/exec_cia.c]: I can do that.

Comment (meepingsnesroms) [reply ctr/exec-3dsx/exec_cia.c]: I am sorry, its actually way to much extra work, and for no benefit, that isnt a speed intensive part of the program, its fine if it does one unnecessary operation.

It just makes the library more messy to add another argument format when this is what most people would want it to be anyway.

Comment (aliaspider) [reply ctr/exec-3dsx/exec_cia.c]: this isn't a debate. the way it is right now is not acceptable.

File: frontend/frontend_driver.c [hunk 5]
```diff
@@ -147,7 +151,10 @@ bool frontend_driver_get_core_extension(char *s, size_t len)
    strlcpy(s, "elf", len);
    return true;
 #elif defined(_3DS)
-   strlcpy(s, "core", len);
+   if (envIsHomebrew())
+      strlcpy(s, "3dsx", len);
+   else
+	  strlcpy(s, "cia", len);
    return true;
```
Comment (aliaspider) [inline frontend/frontend_driver.c]: nit: indent.

File: ctr/exec-3dsx/exec_cia.c (hunk 4 above)
Comment (meepingsnesroms) [reply ctr/exec-3dsx/exec_cia.c]: Its a nitpick, the library is meant to be it own thing, I am not changing it for this one edge case, if you have to don merge it, every time I try to submit something this happens over something irrelevant, its quotes or leave it crashing on launch like before, I am done wasting my time on this nonsense.

**Comment by meepingsnesroms** [comment]
What he is asking is basically to rewrite the entire exec_* functions arg handling, they where made to use the terminal style, it would be way more messy to do it his way then how it is now.

This is just a difference of opinion like tabs vs spaces, its terminal style vs byte array.

**Comment by aliaspider** [comment]
so, let me see this straight, you took the original code, which directly copied the null terminated strings after each other, and changed it to something more complex that creates a single a single string with all args in it separated by quotes, just so you can call a function directly after that that parses said string to remove those quotes, because, ... of a difference of opinion?
and then you refuse to correct this because it is a waste of your time? ok I see.

**Comment by meepingsnesroms** [comment]
You want me to change the library to match this one circumstance, it is meant to be a general 3dsx/cia execution library, not just part of RetroArch.

Why does it matter so much to you that I pass the args as single string instead of the old way?
Why is the old way the correct way?

**Comment by aliaspider** [comment]
because it doesn't need parsing, duh.
do I have to tell you everything ? you made a library with a bad interface, instead of correcting it, you just keep discussing this.
nobody passes quoted args in C, there are only null terminated strings, either a count and a pointer to the first one, or a null terminated char* array.

**Comment by meepingsnesroms** [comment]
Ok I fixed it properly, It actually is cleaner because I could remover the parser entirely.

**Comment by aliaspider** [comment]
much better !
I think it can be merged now.

**Comment by twinaphex** [comment]
Alright, this is great to see how things turned out around here and how the peace was maintained. Much thanks to both you guys.

**Comment by Ferk** [comment]
Awesome @meepingsnesroms, thanks for your work maintaining the 3ds port
